### PR TITLE
Ostapchuk/filtration

### PR DESCRIPTION
**Pull request by YehorOstapchuk**

### CODE REVIEW

**Review by v-ivanchuk — APPROVED** [review]

File: OutOfSchool/OutOfSchool.WebApi/Services/CategoryService.cs [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        {
+            logger.Information("Category creating was started");
+
+            var category = dto.ToDomain();
+
+            CategoryValidation(dto);
+
+            var newCategory = await repository.Create(category).ConfigureAwait(false);
+
+            logger.Information("Category created successfully.");
```
Comment (DmyMi) [inline OutOfSchool/OutOfSchool.WebApi/Services/CategoryService.cs:48]: maybe include `id` of the category you created?

**Review by DmyMi — APPROVED** [review]

File: OutOfSchool/OutOfSchool.DataAccess/Models/Category.cs [hunk 2]
```diff
@@ -1,12 +1,20 @@
...
+using System.ComponentModel.DataAnnotations;
 
 namespace OutOfSchool.Services.Models
 {
     public class Category
     {
         public long Id { get; set; }
 
-        public string Title { get; set; } = string.Empty;
+        [Required(ErrorMessage = "Title is required")]
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.DataAccess/Models/Category.cs:10]: should we move to resources?

File: OutOfSchool/OutOfSchool.WebApi.Tests/Services/ChildServiceTests.cs [hunk 3]
```diff
@@ -19,6 +19,8 @@ public class ChildServiceTests
         private Mock<ILogger> logger;
         private Mock<IStringLocalizer<SharedResource>> localizer;
         private Mock<IEntityRepository<Child>> mockRepository;
+        private Child createdChild = new Child { Id = 14, FirstName = "fn4", LastName = "ln4", Patronymic = "mn4", DateOfBirth = new DateTime(2006, 11, 2), Gender = Gender.Male, ParentId = 1, SocialGroupId = 1 };
+        private ChildDTO createdChildDTO = new ChildDTO { Id = 14, FirstName = "fn4", LastName = "ln4", Patronymic = "mn4", DateOfBirth = new DateTime(2006, 11, 2), Gender = Gender.Male, ParentId = 1, SocialGroupId = 1 };
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi.Tests/Services/ChildServiceTests.cs:23]: please check this failure
'ChildDTO' does not contain a definition for 'Patronymic'

File: OutOfSchool/OutOfSchool.WebApi/Services/SubcategoryService.cs [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+            logger.Information($"Successfuly got a subcategory with id = {id}.");
+
+            return category.ToModel();
+        }
+
+        /// <inheritdoc/>
+        public async Task<IEnumerable<SubcategoryDTO>> GetByCategoryId(long id)
+        {
+            logger.Information("Process of getting all Subcategories started.");
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi/Services/SubcategoryService.cs:109]: maybe just 'getting all Subcategories started'
?

File: OutOfSchool/OutOfSchool.WebApi/Services/SubcategoryService.cs [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+                ? "Subcategory table is empty."
+                : "Successfully got all records from the Subcategory table.");
+
+            return categories.Select(parent => parent.ToModel()).ToList();
+        }
+
+        /// <inheritdoc/>
+        public async Task<SubcategoryDTO> Update(SubcategoryDTO dto)
+        {
+            logger.Information("Subcategory updating was launched.");
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi/Services/SubcategoryService.cs:125]: maybe launched->started
to be consisted?

File: OutOfSchool/OutOfSchool.WebApi/Services/SubcategoryService.cs [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+            {
+                var category = await repository.Update(dto.ToDomain()).ConfigureAwait(false);
+
+                logger.Information("Subcategory succesfully updated.");
+
+                return category.ToModel();
+            }
+            catch (DbUpdateConcurrencyException)
+            {
+                logger.Error("Updating failed. There is no subcategory in the Db with such an id.");
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi/Services/SubcategoryService.cs:139]: with such id
without 'an'

File: OutOfSchool/OutOfSchool.WebApi/Services/SubcategoryService.cs [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+            }
+
+            if (!repository.CategoryExists(dto.CategoryId))
+            {
+                throw new ArgumentException(localizer["There is no category with such id."]);
+            }
+
+            if (repository.SameExists(dto.ToDomain()))
+            {
+                throw new ArgumentException(localizer["There is already a subcategory with such a data."]);
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi/Services/SubcategoryService.cs:158]: with such data

File: OutOfSchool/OutOfSchool.WebApi/Services/SubsubcategoryService.cs [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+using OutOfSchool.Services.Models;
+using OutOfSchool.Services.Repository;
+using OutOfSchool.WebApi.Extensions;
+using OutOfSchool.WebApi.Models;
+using Serilog;
+
+namespace OutOfSchool.WebApi.Services
+{
+    /// <summary>
+    /// Service with business logic for SubsubcategoryController.
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi/Services/SubsubcategoryService.cs:16]: maybe just 'Subsubcategory Service' ?
or put some high level details about what this service does
but no reference to controller, it is not needed

File: OutOfSchool/OutOfSchool.WebApi/Services/SubsubcategoryService.cs [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+
+            logger.Information("Subsubcategory created successfully.");
+
+            return newCategory.ToModel();
+        }
+
+        /// <inheritdoc/>
+        public async Task Delete(long id)
+        {
+            logger.Information("Subsubcategoty deleting was launched.");
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi/Services/SubsubcategoryService.cs:56]: started or launched?

File: OutOfSchool/OutOfSchool.WebApi/Services/WorkshopService.cs [hunk 10]
```diff
@@ -125,5 +129,99 @@ public async Task Delete(long id)
...
+            }
+        }
+
+        private Expression<Func<Workshop, bool>> PredicateBuild(WorkshopFilter filter)
+        {
+            var predicate = PredicateBuilder.True<Workshop>();
+
+            if (!string.IsNullOrEmpty(filter.SearchFieldText))
+            {
+                predicate = predicate.And(x => x.Title.Contains(filter.SearchFieldText));
```
Comment (dmitrykiev) [inline OutOfSchool/OutOfSchool.WebApi/Services/WorkshopService.cs:169]: please address the warning below

**Pull request closed without merging**